docs/neural_modeling/plot_05_place_cells.py

```diff
@@ -0,0 +1,410 @@
+}
+features = {
+    "position": (position,),
+    "position + speed": (position, speed),
+    "position + phase": (position, theta),
+    "position + phase + speed": (position, theta, speed),
+}
+
+# %%
+# In a loop, we can (1) evaluate the basis, (2), fit the model, (3) compute the score and (4) predict the firing rate. For evaluating the score, we can define a train set of intervals and a test set of intervals.
+
+train_iset = position.time_support[
+    0 : len(position.time_support) // 2
+]  # Taking the first half of the recording as the train set
+test_iset = position.time_support[
+    len(position.time_support) // 2 :
+]  # Taking the second half of the recording as the test set
+
+# %%
+# Let's train all the models.
+scores = {}
+predicted_rates = {}
+
+for m in models:
+    print("1. Evaluating basis : ", m)
+    X = models[m].evaluate(*features[m])
+
+    print("2. Fitting model : ", m)
+    # glm = nmo.glm.GLM()
+    glm.fit(
+        X.restrict(train_iset)[:, np.newaxis, :],
+        count.restrict(train_iset)[:, np.newaxis],
+    )
+
+    print("3. Scoring model : ", m)
+    scores[m] = glm.score(
+        X.restrict(test_iset)[:, np.newaxis, :],
+        count.restrict(test_iset)[:, np.newaxis],
+        score_type="pseudo-r2-McFadden",
+    )
+
+    print("4. Predicting rate")
+    predicted_rates[m] = glm.predict(X.restrict(test_iset)[:, np.newaxis, :]) / bin_size
+
+
+scores = pd.Series(scores)
+scores = scores.sort_values()
+
+# %%
+# Let's compute scores for each models.
+
+plt.figure(figsize=(5, 3))
+plt.barh(np.arange(len(scores)), scores)
+plt.yticks(np.arange(len(scores)), scores.index)
+plt.ylabel("Pseudo r2")
+plt.tight_layout()
+
+
+# %%
+# Some models are clearly doing better than others.
+#
+# !!! warning " A proper model comparison should be done by scoring models repetitively on various train and test set. Here we are only doing partial models comparison for the sake of conciseness. "
+#
+# Alternatively, we can plot some tuning curves to compare each models visually.
+
+tuning_curves = {}
+
+for m in models:
+    tuning_curves[m] = {
+        "position": nap.compute_1d_tuning_curves_continuous(
+            predicted_rates[m], position, 50
+        ),
+        "speed": nap.compute_1d_tuning_curves_continuous(
+            predicted_rates[m], speed, 30, minmax=(0, 100)
+        ),
+    }
+
+
+def plot_tuning_curves(axis, tc, m):
+    gs = axis.subgridspec(1, 2)
+    plt.subplot(gs[0, 0])
+    plt.plot(pf[neuron], "--", label="Observed")
+    plt.plot(tc["position"][0])
+    plt.xlabel("Position (cm)")
+    plt.ylabel("Firing rate (Hz)")
+    plt.title("Model : {}".format(m))
+    plt.legend()
+
+    plt.subplot(gs[0, 1])
+    plt.plot(tc_speed[neuron], "--")
+    plt.plot(tc["speed"][0])
+    plt.xlabel("Speed (cm/s)")
+
+
+fig = plt.figure(figsize=(8, 4))
+outer_grid = fig.add_gridspec(2, 2)
+for i, m in enumerate(models):
+    plot_tuning_curves(outer_grid[i // 2, i % 2], tuning_curves[m], m)
```

Comment:
Fits looks poor, I think it is because you did not restrict to the right interval set before computing the tuning curves from the raw spikes. I am pretty sure it would look much better if you did. We should restrict before fitting and recompute tuning before fitting. 

Another way could be computing the firing rate over the whole recording, and then compute the tuning curve on the desired iset for both spikes and rates at the same time.

Like in your loop
```python
for m in models:
    print("1. Evaluating basis : ", m)
    X = models[m].evaluate(*features[m])
    
    # ... rest of the loop
    # ...

    print("4. Predicting rate")
    # DO NOT RESTRICT HERE
    predicted_rates[m] = glm.predict(X[:, np.newaxis, :]) / bin_size
```

After the loop you can compute the tuning passing the epoch of the test set

```python

 tuning_curves_raw = {
        "position": nap.compute_1d_tuning_curves(
            spikes, position, 50, ep=test_iset
        ),
        "speed": nap.compute_1d_tuning_curves_continuous(
            predicted_rates[m], speed, 30, minmax=(0, 100), ep=test_iset
        ),
    }

tuning_curves_model = {}

for m in models:
    tuning_curves_model[m] = {
        "position": nap.compute_1d_tuning_curves_continuous(
            predicted_rates[m], position, 50, ep=test_iset
        ),
        "speed": nap.compute_1d_tuning_curves_continuous(
            predicted_rates[m], speed, 30, minmax=(0, 100), ep=test_iset
        ),
    }
  
```

Reply:
Done